File: tui/src/uninstall.rs
use anyhow::Result;
use std::env;
use std::fs;
use std::io::{IsTerminal, Write};
use std::path::PathBuf;

/// Helper to determine if a line in a shell profile file is related to RTB.
pub fn is_rtb_profile_line(line: &str) -> bool {
    let l = line.trim();
    if l.contains("rtb shell-init")
        || l.contains("rtb.exe shell-init")
        || l.contains("rtb _goto-resolve")
    {
        return true;
    }
    if l.starts_with("# RTB") || l.starts_with("# rtb") {
        return true;
    }
    let lower = l.to_lowercase();
    if lower.contains("import-module") {
        if lower.contains("rtb")
            || lower.contains("dev-tools")
            || lower.contains("dev-cli")
            || lower.contains("rtb-command-tool")
        {
            return true;
        }
    }
    false
}

/// Helper to determine if a line in a shell profile sets PATH for RTB.
pub fn is_rtb_path_line(line: &str) -> bool {
    let lower = line.trim().to_lowercase();
    (lower.contains("path") || lower.contains("fish_add_path"))
        && (lower.contains("rtb/bin") || lower.contains("rtb\\bin") || lower.contains(".config/rtb"))
}

/// Resolves standard shell profile file paths across operating systems.
pub fn get_profile_paths() -> Vec<PathBuf> {
    if let Ok(override_paths) = env::var("RTB_TEST_PROFILE_PATHS") {
        return override_paths
            .split(';')
            .filter(|s| !s.is_empty())
            .map(PathBuf::from)
            .collect();
    }

    let mut paths = Vec::new();

    if let Ok(prof) = env::var("PROFILE") {
        if !prof.is_empty() {
            paths.push(PathBuf::from(prof));
        }
    }

    if let Some(home) = dirs::home_dir() {
        // Windows PowerShell profile paths
        #[cfg(target_os = "windows")]
        {
            if let Some(docs) = dirs::document_dir() {
                paths.push(docs.join("PowerShell").join("Microsoft.PowerShell_profile.ps1"));
                paths.push(docs.join("WindowsPowerShell").join("Microsoft.PowerShell_profile.ps1"));
                paths.push(docs.join("PowerShell").join("Profile.ps1"));
            }
            paths.push(home.join("Documents").join("PowerShell").join("Microsoft.PowerShell_profile.ps1"));
            paths.push(home.join("Documents").join("WindowsPowerShell").join("Microsoft.PowerShell_profile.ps1"));
        }

        // Unix shell RC paths
        paths.push(home.join(".bashrc"));
        paths.push(home.join(".bash_profile"));
        paths.push(home.join(".zshrc"));
        paths.push(home.join(".profile"));
        paths.push(home.join(".config").join("fish").join("config.fish"));
    }

    paths.sort();
    paths.dedup();
    paths
}

/// Strip RTB shell init lines, legacy Phase 2 imports, and PATH lines from shell profiles.
pub fn clean_shell_profiles() -> Result<()> {
    for profile_path in get_profile_paths() {
        if !profile_path.is_file() {
            continue;
        }

        let content = match fs::read_to_string(&profile_path) {
            Ok(c) => c,
            Err(_) => continue,
        };

        let lines: Vec<&str> = content.lines().collect();
        let filtered: Vec<&str> = lines
            .into_iter()
            .filter(|l| !is_rtb_profile_line(l) && !is_rtb_path_line(l))
            .collect();

        let new_content = if filtered.is_empty() {
            String::new()
        } else {
            let mut joined = filtered.join("\n");
            if content.ends_with('\n') || content.ends_with("\r\n") {
                joined.push('\n');
            }
            joined
        };

        if new_content != content {
            let _ = fs::write(&profile_path, new_content);
        }
    }
    Ok(())
}

/// Remove rtb binary from system bin directory.
pub fn remove_binary() -> Result<()> {
    let mut candidate_bins: Vec<PathBuf> = Vec::new();

    if let Ok(test_bin_dir) = env::var("RTB_TEST_BIN_DIR") {
        let p = PathBuf::from(test_bin_dir);
        candidate_bins.push(p.join("rtb.exe"));
        candidate_bins.push(p.join("rtb"));
    } else if let Ok(bin_dir) = env::var("RTB_BIN_DIR") {
        let p = PathBuf::from(bin_dir);
        candidate_bins.push(p.join("rtb.exe"));
        candidate_bins.push(p.join("rtb"));
    } else {
        if let Some(config_dir) = dirs::config_dir() {
            candidate_bins.push(config_dir.join("rtb").join("bin").join("rtb.exe"));
            candidate_bins.push(config_dir.join("rtb").join("bin").join("rtb"));
        }
        if let Some(home) = dirs::home_dir() {
            candidate_bins.push(home.join(".config").join("rtb").join("bin").join("rtb.exe"));
            candidate_bins.push(home.join(".config").join("rtb").join("bin").join("rtb"));
            candidate_bins.push(home.join(".local").join("bin").join("rtb.exe"));
            candidate_bins.push(home.join(".local").join("bin").join("rtb"));
        }
    }

    for bin_path in &candidate_bins {
        if bin_path.is_file() {
            let _ = fs::remove_file(bin_path);
            if let Some(parent) = bin_path.parent() {
                if parent.file_name().map_or(false, |n| n == "bin") {
                    let _ = fs::remove_dir(parent);
                }
            }
        }
    }

    if env::var("RTB_MOCK_DO_NOT_REMOVE_EXE").is_err() {
        if let Ok(exe_path) = env::current_exe() {
            if let Some(stem) = exe_path.file_stem() {
                if stem.to_string_lossy().eq_ignore_ascii_case("rtb") {
                    let _ = fs::remove_file(&exe_path);
                }
            }
        }
    }

    Ok(())
}

/// Remove user configuration file and directory.
pub fn remove_config() -> Result<()> {
    if let Ok(test_config_path) = env::var("RTB_TEST_CONFIG_PATH") {
        let p = PathBuf::from(test_config_path);
        if p.is_file() {
            let _ = fs::remove_file(&p);
        }
        if let Some(parent) = p.parent() {
            if parent.exists() && parent.file_name().map_or(false, |n| n == "rtb") {
                let _ = fs::remove_dir_all(parent);
            }
        }
        return Ok(());
    }

    if let Some(config_dir) = dirs::config_dir() {
        let rtb_dir = config_dir.join("rtb");
        if rtb_dir.exists() {
            let _ = fs::remove_dir_all(&rtb_dir);
        }
    }

    if let Some(home) = dirs::home_dir() {
        let rtb_dir = home.join(".config").join("rtb");
        if rtb_dir.exists() {
            let _ = fs::remove_dir_all(&rtb_dir);
        }
    }

    Ok(())
}

/// Remove rtb bin path entry from environment / registry.
pub fn clean_path_entry() -> Result<()> {
    if env::var("RTB_TEST_SKIP_PATH_REMOVAL").is_ok() {
        return Ok(());
    }

    #[cfg(target_os = "windows")]
    {
        let script = r#"
            $userPath = [Environment]::GetEnvironmentVariable('Path', 'User')
            if ($userPath) {
                $newPath = ($userPath -split ';' | Where-Object { $_ -and $_ -notlike '*\rtb\bin*' -and $_ -notlike '*\.config\rtb\bin*' }) -join ';'
                [Environment]::SetEnvironmentVariable('Path', $newPath, 'User')
            }
        "#;
        let _ = std::process::Command::new("powershell")
            .args(["-NoProfile", "-NonInteractive", "-Command", script])
            .output();
    }

    Ok(())
}

/// Main execution function for `rtb uninstall`.
pub fn execute_uninstall(yes: bool) -> Result<i32> {
    let is_non_interactive = yes
        || env::var("RTB_NON_INTERACTIVE").is_ok()
        || env::var("CI").is_ok()
        || env::var("GITHUB_ACTIONS").is_ok()
        || !std::io::stdin().is_terminal();

    if !is_non_interactive {
        print!("Are you sure you want to uninstall RTB from your system? (y/N) ");
        std::io::stdout().flush()?;

        let mut input = String::new();
        std::io::stdin().read_line(&mut input)?;
        let trimmed = input.trim().to_lowercase();
        if trimmed != "y" && trimmed != "yes" {
            println!("Uninstallation canceled.");
            return Ok(0);
        }
    }

    remove_binary()?;
    remove_config()?;
    clean_shell_profiles()?;
    clean_path_entry()?;

    println!("Uninstallation Complete! RTB has been removed from your system.");
    Ok(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_rtb_profile_line() {
        assert!(is_rtb_profile_line("Invoke-Expression (& rtb shell-init pwsh)"));
        assert!(is_rtb_profile_line("eval \"$(rtb shell-init bash)\""));
        assert!(is_rtb_profile_line("rtb _goto-resolve myproject"));
        assert!(is_rtb_profile_line("# RTB Autoload"));
        assert!(is_rtb_profile_line("Import-Module \"C:\\Users\\dev\\rtb\\module\\rtb.psd1\""));
        assert!(is_rtb_profile_line("Import-Module rtb"));

        assert!(!is_rtb_profile_line("alias ll='ls -la'"));
        assert!(!is_rtb_profile_line("Import-Module PSReadLine"));
        assert!(!is_rtb_profile_line("export PATH=$PATH:/usr/local/bin"));
    }

    #[test]
    fn test_is_rtb_path_line() {
        assert!(is_rtb_path_line("export PATH=\"$HOME/.config/rtb/bin:$PATH\""));
        assert!(is_rtb_path_line("fish_add_path /home/user/.config/rtb/bin"));
        assert!(!is_rtb_path_line("export PATH=\"$HOME/bin:$PATH\""));
    }
}

File: core/src/commands/uninstall.ts
import type { Command } from 'commander';
import chalk from 'chalk';
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import readline from 'node:readline';
import { spawnSync } from 'node:child_process';
import type { CliContext } from '../types/context.js';
import { outputError, outputJson } from '../utils/output.js';

export function cleanProfileContent(content: string): string {
  const lines = content.split(/\r?\n/);
  const filtered = lines.filter((line) => {
    if (/rtb\s+shell-init/i.test(line)) return false;
    if (/#\s*RTB\s+Shell\s+Integration/i.test(line)) return false;
    if (/Import-Module.*?(rtb|dev-tools|dev-cli|rtb-command-tool).*?\.psd1/i.test(line)) return false;
    if (/#\s*RTB.*?Module/i.test(line)) return false;
    return true;
  });

  const sep = content.includes('\r\n') ? '\r\n' : '\n';
  return filtered.join(sep);
}

export function cleanShellProfiles(customCandidates?: string[]): string[] {
  const cleaned: string[] = [];
  const homeDir = os.homedir();

  const candidates: string[] = customCandidates ?? [];
  if (!customCandidates) {
    if (process.platform === 'win32') {
      const docs = path.join(homeDir, 'Documents');
      candidates.push(
        path.join(docs, 'PowerShell', 'Microsoft.PowerShell_profile.ps1'),
        path.join(docs, 'WindowsPowerShell', 'Microsoft.PowerShell_profile.ps1')
      );
    }

    candidates.push(
      path.join(homeDir, '.bashrc'),
      path.join(homeDir, '.bash_profile'),
      path.join(homeDir, '.zshrc'),
      path.join(homeDir, '.config', 'fish', 'config.fish')
    );
  }

  for (const prof of candidates) {
    if (fs.existsSync(prof)) {
      try {
        const content = fs.readFileSync(prof, 'utf-8');
        const newContent = cleanProfileContent(content);

        if (newContent !== content) {
          fs.writeFileSync(prof, newContent, 'utf-8');
          cleaned.push(prof);
        }
      } catch {}
    }
  }

  return cleaned;
}

export function findRtbExecutables(): string[] {
  const executables: string[] = [];
  const candidateDirs: string[] = [];

  if (process.platform === 'win32') {
    if (fs.existsSync('D:\\bin')) {
      candidateDirs.push('D:\\bin');
    }
    if (process.env.LOCALAPPDATA) {
      candidateDirs.push(path.join(process.env.LOCALAPPDATA, 'Programs', 'rtb', 'bin'));
    }
  }

  // Probe via where.exe / which
  try {
    const isWindows = process.platform === 'win32';
    const cmd = isWindows ? 'where.exe' : 'which';
    for (const binName of ['rtb', 'rtbtui']) {
      const binArgs = isWindows ? [binName] : ['-a', binName];
      const res = spawnSync(cmd, binArgs, { encoding: 'utf-8' });
      if (res.status === 0 && res.stdout) {
        const found = res.stdout.split(/\r?\n/).map((l) => l.trim()).filter(Boolean);
        for (const f of found) {
          const dir = path.dirname(f);
          if (!candidateDirs.includes(dir)) {
            candidateDirs.push(dir);
          }
        }
      }
    }
  } catch {}

  const targetFilenames = ['rtb', 'rtb.cmd', 'rtb.ps1', 'rtb.js', 'rtbtui', 'rtbtui.exe'];

  for (const dir of candidateDirs) {
    for (const file of targetFilenames) {
      const fullPath = path.join(dir, file);
      if (fs.existsSync(fullPath)) {
        const norm = path.normalize(fullPath).toLowerCase();
        // Prevent deleting source or build files inside the development repository
        if (
          !norm.includes(path.normalize('rtb-command-tool').toLowerCase()) &&
          !executables.includes(fullPath)
        ) {
          executables.push(fullPath);
        }
      }
    }
  }

  return executables;
}

export function performUninstall(options: { keepConfig?: boolean; customConfigDir?: string } = {}): {
  removedPaths: string[];
  cleanedProfiles: string[];
} {
  const removedPaths: string[] = [];
  const homeDir = os.homedir();
  const userConfigDir = options.customConfigDir || path.join(homeDir, '.config', 'rtb');
  const binDir = process.env.RTB_BIN_DIR || path.join(userConfigDir, 'bin');

  // 1. Clean shell profiles
  const cleanedProfiles = cleanShellProfiles();

  // 2. Remove binary directory
  if (fs.existsSync(binDir)) {
    try {
      fs.rmSync(binDir, { recursive: true, force: true });
      removedPaths.push(binDir);
    } catch {}
  }

  // 2b. Remove discovered standalone binary wrappers from PATH (e.g. D:\bin, etc.)
  const discoveredExecutables = findRtbExecutables();
  for (const exe of discoveredExecutables) {
    try {
      if (fs.existsSync(exe)) {
        fs.unlinkSync(exe);
        removedPaths.push(exe);
      }
    } catch {}
  }

  // 2c. Check and uninstall global npm package if present
  try {
    const isWindows = process.platform === 'win32';
    const npmCmd = isWindows ? 'npm.cmd' : 'npm';
    spawnSync(npmCmd, ['uninstall', '-g', '@3mr5aled/rtb', '@3mr-5aled/rtb'], { stdio: 'ignore' });
  } catch {}

  // 3. Remove user configuration directory if not keepConfig
  if (!options.keepConfig && fs.existsSync(userConfigDir)) {
    const normConfigDir = path.normalize(userConfigDir).toLowerCase();
    const isInsideRepo = normConfigDir.includes(path.normalize('rtb-command-tool').toLowerCase());
    const isTargetConfig =
      path.basename(userConfigDir) === 'rtb' ||
      normConfigDir.endsWith(path.normalize('.config/rtb').toLowerCase()) ||
      normConfigDir.includes(path.normalize('rtb-uninstall-test').toLowerCase());

    if (!isInsideRepo && isTargetConfig) {
      try {
        fs.rmSync(userConfigDir, { recursive: true, force: true });
        removedPaths.push(userConfigDir);
      } catch {}
    } else if (!isInsideRepo && fs.existsSync(path.join(userConfigDir, 'rtb.config.json'))) {
      try {
        fs.unlinkSync(path.join(userConfigDir, 'rtb.config.json'));
        removedPaths.push(path.join(userConfigDir, 'rtb.config.json'));
      } catch {}
    }
  }

  // 4. Legacy AppData cleanup on Windows
  if (process.platform === 'win32' && process.env.APPDATA) {
    const legacyRoaming = path.join(process.env.APPDATA, 'rtb');
    if (fs.existsSync(legacyRoaming)) {
      try {
        fs.rmSync(legacyRoaming, { recursive: true, force: true });
        removedPaths.push(legacyRoaming);
      } catch {}
    }
  }

  // 5. Clean PATH on Windows
  if (process.platform === 'win32') {
    try {
      const cleanPs = `
        $cur = [Environment]::GetEnvironmentVariable('PATH', 'User')
        if ($cur) {
          $parts = @($cur -split ';' | Where-Object { $_ -and $_ -notmatch '(?i)\\.config[\\\\/]rtb[\\\\/]bin' -and $_ -notmatch '(?i)AppData[\\\\/]Roaming[\\\\/]rtb' })
          [Environment]::SetEnvironmentVariable('PATH', ($parts -join ';'), 'User')
        }
      `;
      spawnSync('powershell.exe', ['-NoProfile', '-Command', cleanPs], { stdio: 'ignore' });
    } catch {}
  }

  return { removedPaths, cleanedProfiles };
}

export function registerUninstallCommand(program: Command, getContext: () => CliContext): void {
  program
    .command('uninstall')
    .description('Cleanly uninstall RTB from system, removing binaries, profile hooks, and config')
    .option('-f, --force', 'Skip interactive confirmation prompt', false)
    .option('--keep-config', 'Preserve workspace configuration and projects', false)
    .action(async (options: { force?: boolean; keepConfig?: boolean }) => {
      const ctx = getContext();

      if (!options.force) {
        if (!ctx.isInteractive) {
          outputError(
            "Uninstallation requires confirmation. Run 'rtb uninstall --force' to proceed non-interactively.",
            'CONFIRMATION_REQUIRED',
            ctx.isJson
          );
          process.exitCode = 1;
          return;
        }

        const rl = readline.createInterface({
          input: process.stdin,
          output: process.stdout,
        });

        const answer: string = await new Promise((resolve) => {
          rl.question(chalk.yellow('  Are you sure you want to uninstall RTB from your system? (y/N) '), (ans) => {
            rl.close();
            resolve(ans.trim().toLowerCase());
          });
        });

        if (answer !== 'y' && answer !== 'yes') {
          console.log(chalk.gray('\n  Uninstallation canceled.\n'));
          return;
        }
      }

      const customConfigDir = ctx.configPath ? path.dirname(ctx.configPath) : undefined;
      const result = performUninstall({
        keepConfig: options.keepConfig,
        customConfigDir,
      });

      if (ctx.isJson) {
        outputJson({
          uninstalled: true,
          removedPaths: result.removedPaths,
          cleanedProfiles: result.cleanedProfiles,
          keptConfig: Boolean(options.keepConfig),
        });
        return;
      }

      console.log(`\n${chalk.cyan('══════════════════════════════════════════')}`);
      console.log(`  ${chalk.bold('rtb (ﺐﺗر) » Uninstallation')}`);
      console.log(`${chalk.cyan('══════════════════════════════════════════')}\n`);

      if (result.removedPaths.length > 0) {
        for (const p of result.removedPaths) {
          console.log(`  ${chalk.green('✓')} Removed: ${chalk.gray(p)}`);
        }
      }

      if (result.cleanedProfiles.length > 0) {
        for (const prof of result.cleanedProfiles) {
          console.log(`  ${chalk.green('✓')} Cleaned profile: ${chalk.gray(prof)}`);
        }
      }

      if (options.keepConfig) {
        console.log(`  ${chalk.yellow('ℹ')} Configuration directory preserved.`);
      }

      console.log(`\n  ${chalk.green('✓ RTB has been successfully uninstalled from your system.')}\n`);
    });
}
